util: Add Assume() identity function

maflcko opened the pull request:
This is needed for #20138. Please refer to the added documentation for motivation.

practicalswift committed:
util: Allow Assert(...) to be used in all contexts

Fixes the compile error when used inside operator[]:

./chain.h:404:23: error: C++11 only allows consecutive left square brackets when introducing an attribute
        return (*this)[Assert(pindex)->nHeight] == pindex;
                      ^

Commit:
util: Add Assume() identity function

laanwj commented:
Concept ACK. It's mildly interesting that we're sort of reinventing the original `assert()` here, which gets disabled on `NDEBUG`, which we had to always force on, but I think it's fine to reinvent the wheel here instead of assuming any specific compiler semantics.

MarcoFalke commented:
Indeed, this `Assume` is like the traditional `assert`.

jnewbery commented:
Concept ACK. If you're willing to entertain paint color requests, I prefer the name `Check()`, as was originally proposed in #16136.

MarcoFalke commented:
I think @ryanofsky also likes `check`, so I might change to that

jnewbery commented:
I think Russ doesn't care about the name, but would prefer a pair of check/dcheck macros: https://github.com/bitcoin/bitcoin/pull/16136#issuecomment-545993867

sipa commented:
FWIW, libsecp256k1 has CHECK() (always enabled) and VERIFY_CHECK() functions (enabled with -DVERIFY, used in test builds).

practicalswift commented:
Concept ACK

MarcoFalke commented:
> FWIW, libsecp256k1 has CHECK() (always enabled) and VERIFY_CHECK() functions (enabled with -DVERIFY, used in test builds).

That is good to know, because that means we can't use that name due to name-clashing.

ajtowns reviewed (commented):
ACK fa9672603560320bc9f1017a7de5e85de10d9561

I think this is a good choice of behaviour that should be easy to use correctly. Would like to see the ability to turn on logging of violated Assumptions even in production builds in a followup.

practicalswift commented:
ACK fa9672603560320bc9f1017a7de5e85de10d9561: patch looks good

hebasto reviewed (commented):
Approach ACK fa9672603560320bc9f1017a7de5e85de10d9561

Are there any cases when `assert` is preferable to `Assert`. If "yes", it would be nice to have an example in the docs. If "no", why do we mention `assert` in the docs?

Commit:
util: Remove probably misleading TODO

The TODO has been added by me, but I don't remember how to solve it. The
current code works fine, so just remove the TODO.

MarcoFalke commented on doc/developer-notes.md:
```diff
@@ -276,6 +276,31 @@ configure option adds `-DDEBUG_LOCKORDER` to the compiler flags. This inserts
+     logic bug can be shown to the user and the program can continue to run.
+* `Assume` can be used for nonfatal internal logic bugs. In debug builds it
+  behaves like Assert()/assert() to notify developers and testers about
+  non-fatal errors. In production it doesn't warn or log anything.
```
thx, done

MarcoFalke commented on doc/developer-notes.md:
```diff
@@ -276,6 +276,31 @@ configure option adds `-DDEBUG_LOCKORDER` to the compiler flags. This inserts
+   - For example it can be assumed that a variable is only initialized once,
+     but a failed assumption does not result in a fatal bug. A failed
+     assumption may or may not result in a slightly degraded user experience,
+     but it is safe to continue program execution.
```
> Maybe rephrase it as a directive

thx, done

> maybe we should deprecate assert()

I'll leave that for a follow-up

MarcoFalke commented:
> Are there any cases when assert is preferable to Assert

No

> If "no", why do we mention assert in the docs?

It documents how the `assert` is used in existing code

practicalswift commented:
cr ACK faa05854f832405231c9198787a4eafe2cd4c5f0

jnewbery commented:
utACK faa05854f832405231c9198787a4eafe2cd4c5f0

jonatack commented on doc/developer-notes.md:
```diff
@@ -276,6 +276,33 @@ configure option adds `-DDEBUG_LOCKORDER` to the compiler flags. This inserts
+  failure, it will throw an exception, which can be caught to recover from the
+  error.
+   - For example, a nullptr dereference or any other logic bug in RPC code
+     means that the RPC code is faulty and can not be executed. However, the
```
s/can not/cannot/

jonatack commented on doc/developer-notes.md:
```diff
@@ -276,6 +276,33 @@ configure option adds `-DDEBUG_LOCKORDER` to the compiler flags. This inserts
+  behaves like `Assert`/`assert` to notify developers and testers about
+  nonfatal errors. In production it doesn't warn or log anything, though the
+  expression is always evaluated.
+   - For example it can be assumed that a variable is only initialized once,
```
s/For example it can/For example, it may/

jonatack reviewed (commented):
Approach ACK, some doc nits if you retouch

MarcoFalke commented:
Will leave style nits alone to preserve ACKs

MarcoFalke commented:
I think this is ready, but I am waiting on one more (re?)ACK before merge

hebasto reviewed (approved):
ACK faa05854f832405231c9198787a4eafe2cd4c5f0, I have reviewed the code and it looks OK, I agree it can be merged.
